src/pornin.rs

```diff
@@ -0,0 +1,415 @@
+            if a & 1 > 0 {
+                if a < b {
+                    (a, b, u, v) = (b, a, v, u);
+                    // In both the aforesaid Pornin's method and its modification "n" and "d"
+                    // could not become negative simultaneously even if they were updated after
+                    // each iteration of the inner loop. Also at this point they both have odd
+                    // values. Therefore, the quadratic reciprocity law for the modified Jacobi
+                    // symbol (x / |y|) can be used. According to it, if both x and y are odd
+                    // numbers, among which there is a positive one, then for x = y = 3 (mod 4)
+                    // we have (x / |y|) = -(y / |x|) and for either x or y equal 1 modulo 4
+                    // the symbols (x / |y|) and (y / |x|) are equal
+                    t ^= a & b;
+                }
+                a = (a - b) >> 1;
+                u = (u.0 - v.0, u.1 - v.1);
+                v = (v.0 << 1, v.1 << 1);
+                // The modified Jacobi symbol (2 / |y|) is -1, iff y mod 8 is {3, 5}
+                t ^= b ^ b >> 1;
+                i -= 1;
+            } else {
+                // Performing the batch of sequential iterations, which divide "a" by 2
+                let z = i.min(a.trailing_zeros());
+                // The modified Jacobi symbol (2 / |y|) is -1, iff y mod 8 is {3, 5}. However,
+                // we do not need its value for a batch with an even number of divisions by 2
+                t ^= (b ^ b >> 1) & (z << 1) as u64;
+                v = (v.0 << z, v.1 << z);
+                a >>= z;
+                i -= z;
+            }
+        }
+        (n, d) = ((&n * u.0 + &d * u.1) >> 30, (&n * v.0 + &d * v.1) >> 30);
+
+        // This fragment is present to guarantee the correct behavior of the function
+        // in the case of arguments, whose greatest common divisor is no less than 2^64
+        if n == LInt::ZERO {
+            // In both the aforesaid Pornin's method and its modification the pair of the values
+            // of "n" and "d" after the divergence point contains a positive number and a negative
+            // one. Since the value of "n" is 0, the divergence point has not been reached by the
+            // inner loop this time, so there is no need to check whether "d" is equal to -1
+            return (d == LInt::ONE) as i64 * (1 - (t & 2) as i64);
+        }
+
+        if n.is_negative() {
+            // Since in both the aforesaid Pornin's method and its modification "d" is always odd
+            // and cannot become negative simultaneously with "n", the value of "d" is positive.
+            // The modified Jacobi symbol (-1 / |y|) for a positive y is -1, iff y mod 4 = 3
+            t ^= d.0[0];
+            n = -n;
+        } else if d.is_negative() {
+            // The modified Jacobi symbols (x / |y|) and (x / |-y|) are equal, so "t" is not updated
+            d = -d;
+        }
+    }
```

Comment:
In the paper there is an extra reduction before exiting the loop
![image](https://github.com/taikoxyz/halo2curves/assets/22738317/4dbb9181-4a08-429e-8c8e-d189ce5a81db)

I may have missed it, but does it appear somewhere?

Reply:
No, this is done for modular inversion. Neither GCD and Jacobi symbol do no require this part.